azure-batch/azure/batch/custom/patch.py

```diff
@@ -0,0 +1,285 @@
+                add_collection_response = add_collection_response.output
+
+            for task_result in add_collection_response.value:  # pylint: disable=no-member
+                if task_result.status == TaskAddStatus.server_error:
+                    # Server error will be retried
+                    with self._pending_queue_lock:
+                        for task in chunk_tasks_to_add:
+                            if task.id == task_result.task_id:
+                                self._tasks_to_add.appendleft(task)
+                elif (task_result.status == TaskAddStatus.client_error
+                        and not task_result.error.code == "TaskExists"):
+                    # Client error will be recorded unless Task already exists
+                    self._failures.appendleft(task_result)
+                else:
+                    results_queue.appendleft(task_result)
+
+    def task_collection_thread_handler(self, results_queue):
+        """Main method for worker to run
+
+        Pops a chunk of tasks off the collection of pending tasks to be added and submits them to be added.
+
+        :param collections.deque results_queue: Queue for worker to output results to
+        """
+        while len(self._tasks_to_add) != 0 and not self._has_early_termination_error:
+            max_tasks = self._max_tasks_per_request  # local copy
+            chunk_tasks_to_add = []
+            with self._pending_queue_lock:
+                while len(chunk_tasks_to_add) < max_tasks and len(self._tasks_to_add) != 0:
+                    chunk_tasks_to_add.append(self._tasks_to_add.pop())
+
+            if len(chunk_tasks_to_add) != 0:
+                self._bulk_add_tasks(results_queue, chunk_tasks_to_add)
+
+
+        # Only define error if all threads have finished and there were failures
+        with self._error_lock:
+            if threading.active_count() == 1 and len(self._failures) > 0:
+                self.error = errors.CreateTasksErrorException(
+                    "One or more tasks failed to be added",
+                    self._failures,
+                    self._tasks_to_add)
+            else:
+                return
+
+def _handle_output(results_queue):
+    """Scan output for exceptions
+
+    If there is an output from an add task collection call add it to the results or throw the respective exception.
+
+    :param results_queue: Queue containing results of attempted add_collection's
+    :type results_queue: collections.deque
+    :return: list of TaskAddResults
+    :rtype: list[~TaskAddResult]
+    """
+    results = []
+    while len(results_queue) != 0:
+        queue_item = results_queue.pop()
+        if isinstance(queue_item, Exception):
+            raise queue_item
+        else:
+            results.append(queue_item)
+    return results
+
+def patch_client(client):
+    batch_client = client
+    def callback_bulk_add_collection(*args, **kwargs):
```

Comment:
Use wraps, otherwise what the IDE sees is the callback, and you loose all intellisense and doc:
https://docs.python.org/3/library/functools.html#functools.wraps

Reply:
You can also uses types to avoid the callback
```python
target.method = types.MethodType(method,target)
```
which use be for you:
```python
client.task.add_collection = types.MethodType(bulk_add_collection, client.task)
```

This would be the prefered approach anyway, since you will loose the `metadata` dict otherwise (won't be copied with the callback)